src/Subscription.js

```diff
@@ -196,6 +214,18 @@ export default class Subscription extends EventEmitter {
             }
             debug('Gap detected, requesting resend for stream %s from %o to %o', this.streamId, from, to)
             this.emit('gap', fromObject, toObject, msg.getPublisherId(), msg.messageId.msgChainId)
+
+            // If for some reason the missing messages are not received, the gap filling request is resent every 'resendTimeout' seconds
+            // until a message is received, at which point the gap will be filled or
+            // a new different gap request will be sent and resent every 'resendTimeout' seconds.
+            clearInterval(this.gaps[key])
+            this.gaps[key] = setInterval(() => {
+                if (this.lastReceivedMsgRef[key].compareTo(to) === -1) {
```

Comment:
Not related to this PR but I do note that items are never removed from `this.lastReceivedMsgRef` so it'll only ever grow in size until the subscription can be garbage collected.

Reply:
Normally we need to remember the reference of the last received message per publisher and message chain id for gap filling / duplicate detection / ordering, so we cannot really delete them. If a publisher is removed, we could delete all related entries (periodically fetch the list of publishers of the stream).
But for the message chain id, we cannot really know if a publisher is ever going to use the same again (unless we receive a "bye" message). A publisher could publish with this chain id only once per day, so if we use a standard expiration time for every key it should be very large. We could also compute the mean `m` and standard deviation `d` of the publication period for each chain id and determine some key-specific expiration time using a linear combination of `m` and `d` like in TCP timeout based on RTT.
A buggy/malicious publisher could also flood messages with a new chain id every time, so there could be a counter for a max number of chain ids per publisher and refuse new chain ids.
Would be good to have @hpihkala and @harbu opinions on this matter and address it in some other PR if this is really an issue.

Reply:
Yep not something for this pr

Reply:
In practice we can assume that the set of `(publisher, msgChainId)` tuples has an upper bound, and client session times are not infinite.

However this does actually open an attack where an evil publisher could rotate their `msgChainId` on every message in an attempt to exhaust memory on subscribed clients. Clients could of course defend by stopping to listen to such streams, so the attack would be kind of feeble.

I don't feel that this needs to be addressed right now in this PR, but @mthambipillai please log a JIRA ticket with minor priority for rethinking whether we could do some kind of evictions from the last received message cache. The evictions can't be based on time for the reasons Melchior mentioned above, but one idea: we could have a hard limit on the number of remembered `msgChainId`s per publisher, which could be set relatively high (e.g. 1000) to defend against the above mentioned attack/publisher malfunction, while allowing for almost all valid cases.

Reply:
Yes the attack you describe and the hard limit solution is what I meant by
> A buggy/malicious publisher could also flood messages with a new chain id every time, so there could be a counter for a max number of chain ids per publisher and refuse new chain ids.

Reply:
Yes, haha, looks like my subconscious read your message more closely than my conscious mind :)

Reply:
👍 for hard limit